File: apps/syn-cli/src/syn_cli/commands/marketplace/_registry.py
"""Marketplace registry management — add, list, remove, refresh."""

from __future__ import annotations

from typing import Annotated

import typer
from rich.table import Table

from syn_cli._output import console, format_timestamp, print_error, print_success
from syn_cli.commands._marketplace_client import (
    fetch_marketplace_json,
    load_registries,
    refresh_index,
    save_cached_index,
    save_registries,
)
from syn_cli.commands._marketplace_models import (
    CachedMarketplace,
    RegistryConfig,
    RegistryEntry,
)

app = typer.Typer(
    name="marketplace",
    help="Manage workflow marketplace registries.",
    no_args_is_help=True,
)


@app.command("add")
def add_marketplace(
    repo: Annotated[
        str,
        typer.Argument(help="GitHub repo (org/repo shorthand)"),
    ],
    ref: Annotated[
        str,
        typer.Option("--ref", "-r", help="Git branch or tag"),
    ] = "main",
    name: Annotated[
        str | None,
        typer.Option("--name", "-n", help="Override registry name"),
    ] = None,
) -> None:
    """Register a GitHub repo as a workflow marketplace."""
    console.print(f"Fetching marketplace.json from [cyan]{repo}[/cyan]@{ref}...")

    try:
        index = fetch_marketplace_json(repo, ref)
    except RuntimeError as e:
        print_error(str(e))
        raise typer.Exit(1) from None

    registry_name = name or index.name

    config = load_registries()
    if registry_name in config.registries:
        print_error(f"Marketplace '{registry_name}' is already registered")
        console.print("[dim]Use 'syn marketplace remove' first to re-register.[/dim]")
        raise typer.Exit(1)

    from datetime import UTC, datetime

    entry = RegistryEntry(
        repo=repo,
        ref=ref,
        added_at=datetime.now(tz=UTC).isoformat(),
    )
    updated = RegistryConfig(
        version=config.version,
        registries={**config.registries, registry_name: entry},
    )
    save_registries(updated)

    # Cache the index we just fetched
    save_cached_index(
        registry_name,
        CachedMarketplace(
            fetched_at=datetime.now(tz=UTC).isoformat(),
            index=index,
        ),
    )

    plugin_count = len(index.plugins)
    print_success(
        f"Added marketplace [bold]{registry_name}[/bold] "
        f"({plugin_count} plugin{'s' if plugin_count != 1 else ''})"
    )


@app.command("list")
def list_marketplaces() -> None:
    """List registered marketplace registries."""
    config = load_registries()

    if not config.registries:
        console.print("[dim]No marketplaces registered.[/dim]")
        console.print(
            "[dim]Add one with: syn marketplace add syntropic137/workflow-library[/dim]"
        )
        return

    table = Table(title="Registered Marketplaces")
    table.add_column("Name", style="bold")
    table.add_column("Repo")
    table.add_column("Ref")
    table.add_column("Added")

    for name, entry in config.registries.items():
        table.add_row(
            name,
            entry.repo,
            entry.ref,
            format_timestamp(entry.added_at),
        )

    console.print(table)


@app.command("remove")
def remove_marketplace(
    name: Annotated[
        str,
        typer.Argument(help="Registry name to remove"),
    ],
) -> None:
    """Remove a registered marketplace."""
    config = load_registries()

    if name not in config.registries:
        print_error(f"Marketplace '{name}' is not registered")
        raise typer.Exit(1)

    remaining = {k: v for k, v in config.registries.items() if k != name}
    updated = RegistryConfig(version=config.version, registries=remaining)
    save_registries(updated)

    # Clean up cache
    from syn_cli.commands._marketplace_client import _CACHE_DIR

    cache_path = _CACHE_DIR / f"{name}.json"
    if cache_path.exists():
        cache_path.unlink()

    print_success(f"Removed marketplace [bold]{name}[/bold]")


@app.command("refresh")
def refresh_marketplace(
    name: Annotated[
        str | None,
        typer.Argument(help="Registry name (refreshes all if omitted)"),
    ] = None,
) -> None:
    """Force-refresh cached marketplace indexes."""
    config = load_registries()

    if not config.registries:
        console.print("[dim]No marketplaces registered.[/dim]")
        return

    targets: list[tuple[str, RegistryEntry]]
    if name:
        if name not in config.registries:
            print_error(f"Marketplace '{name}' is not registered")
            raise typer.Exit(1)
        targets = [(name, config.registries[name])]
    else:
        targets = list(config.registries.items())

    for reg_name, entry in targets:
        console.print(f"Refreshing [cyan]{reg_name}[/cyan]...", end=" ")
        try:
            index = refresh_index(reg_name, entry, force=True)
            plugin_count = len(index.plugins)
            console.print(
                f"[green]done[/green] "
                f"({plugin_count} plugin{'s' if plugin_count != 1 else ''})"
            )
        except RuntimeError as e:
            console.print(f"[red]failed[/red] ({e})")

File: apps/syn-cli-node/src/commands/marketplace/registry.ts
/**
 * Marketplace registry management — add, list, remove, refresh.
 * Port of apps/syn-cli/src/syn_cli/commands/marketplace/_registry.py
 */

import fs from "node:fs";
import path from "node:path";
import type { CommandDef, ParsedArgs } from "../../framework/command.js";
import { CLIError } from "../../framework/errors.js";
import { printError, printSuccess, print, printDim } from "../../output/console.js";
import { style, BOLD, CYAN, GREEN, RED } from "../../output/ansi.js";
import { formatTimestamp } from "../../output/format.js";
import { Table } from "../../output/table.js";
import { synPath } from "../../persistence/store.js";
import {
  fetchMarketplaceJson,
  loadRegistries,
  refreshIndex,
  saveCachedIndex,
  saveRegistries,
  validateRegistryName,
} from "../../marketplace/client.js";
import type { RegistryEntry } from "../../marketplace/models.js";

// ---------------------------------------------------------------------------
// add
// ---------------------------------------------------------------------------

export const addCommand: CommandDef = {
  name: "add",
  description: "Register a GitHub repo as a workflow marketplace",
  args: [{ name: "repo", description: "GitHub repo (org/repo shorthand)", required: true }],
  options: {
    ref: { type: "string", short: "r", description: "Git branch or tag", default: "main" },
    name: { type: "string", short: "n", description: "Override registry name" },
  },
  handler: async (parsed: ParsedArgs) => {
    const repo = parsed.positionals[0];
    if (!repo) {
      printError("Missing required argument: repo");
      throw new CLIError("Missing argument", 1);
    }

    const ref = (parsed.values["ref"] as string | undefined) ?? "main";
    const nameOverride = parsed.values["name"] as string | undefined;

    print(`Fetching marketplace.json from ${style(repo, CYAN)}@${ref}...`);

    let index;
    try {
      index = await fetchMarketplaceJson(repo, ref);
    } catch (err) {
      printError(err instanceof Error ? err.message : String(err));
      throw new CLIError("Fetch failed", 1);
    }

    const registryName = nameOverride ?? index.name;
    try {
      validateRegistryName(registryName);
    } catch (err) {
      printError(err instanceof Error ? err.message : String(err));
      throw new CLIError("Invalid name", 1);
    }

    const config = loadRegistries();
    if (config.registries[registryName]) {
      printError(`Marketplace '${registryName}' is already registered`);
      printDim("Use 'syn marketplace remove' first to re-register.");
      throw new CLIError("Already registered", 1);
    }

    const entry: RegistryEntry = {
      repo,
      ref,
      added_at: new Date().toISOString(),
    };

    saveRegistries({
      version: config.version,
      registries: { ...config.registries, [registryName]: entry },
    });

    saveCachedIndex(registryName, {
      fetched_at: new Date().toISOString(),
      index,
    });

    const pluginCount = index.plugins.length;
    printSuccess(
      `Added marketplace ${style(registryName, BOLD)} ` +
      `(${pluginCount} plugin${pluginCount !== 1 ? "s" : ""})`,
    );
  },
};

// ---------------------------------------------------------------------------
// list
// ---------------------------------------------------------------------------

export const listMarketplaceCommand: CommandDef = {
  name: "list",
  description: "List registered marketplace registries",
  handler: async () => {
    const config = loadRegistries();

    if (Object.keys(config.registries).length === 0) {
      printDim("No marketplaces registered.");
      printDim("Add one with: syn marketplace add syntropic137/workflow-library");
      return;
    }

    const table = new Table({ title: "Registered Marketplaces" });
    table.addColumn("Name", { style: BOLD });
    table.addColumn("Repo");
    table.addColumn("Ref");
    table.addColumn("Added");

    for (const [name, entry] of Object.entries(config.registries)) {
      table.addRow(
        name,
        entry.repo,
        entry.ref,
        formatTimestamp(entry.added_at),
      );
    }
    table.print();
  },
};

// ---------------------------------------------------------------------------
// remove
// ---------------------------------------------------------------------------

export const removeCommand: CommandDef = {
  name: "remove",
  description: "Remove a registered marketplace",
  args: [{ name: "name", description: "Registry name to remove", required: true }],
  handler: async (parsed: ParsedArgs) => {
    const name = parsed.positionals[0];
    if (!name) {
      printError("Missing required argument: name");
      throw new CLIError("Missing argument", 1);
    }

    const config = loadRegistries();
    if (!config.registries[name]) {
      printError(`Marketplace '${name}' is not registered`);
      throw new CLIError("Not found", 1);
    }

    const remaining: Record<string, RegistryEntry> = {};
    for (const [k, v] of Object.entries(config.registries)) {
      if (k !== name) remaining[k] = v;
    }
    saveRegistries({ version: config.version, registries: remaining });

    // Clean up cache
    try {
      validateRegistryName(name);
      const cachePath = path.join(synPath("marketplace", "cache"), `${name}.json`);
      if (fs.existsSync(cachePath)) fs.unlinkSync(cachePath);
    } catch {
      // Skip cleanup for unsafe names
    }

    printSuccess(`Removed marketplace ${style(name, BOLD)}`);
  },
};

// ---------------------------------------------------------------------------
// refresh
// ---------------------------------------------------------------------------

export const refreshCommand: CommandDef = {
  name: "refresh",
  description: "Force-refresh cached marketplace indexes",
  args: [{ name: "name", description: "Registry name (refreshes all if omitted)" }],
  handler: async (parsed: ParsedArgs) => {
    const nameFilter = parsed.positionals[0] as string | undefined;
    const config = loadRegistries();

    if (Object.keys(config.registries).length === 0) {
      printDim("No marketplaces registered.");
      return;
    }

    let targets: [string, RegistryEntry][];
    if (nameFilter) {
      const entry = config.registries[nameFilter];
      if (!entry) {
        printError(`Marketplace '${nameFilter}' is not registered`);
        throw new CLIError("Not found", 1);
      }
      targets = [[nameFilter, entry]];
    } else {
      targets = Object.entries(config.registries);
    }

    for (const [regName, entry] of targets) {
      process.stdout.write(`Refreshing ${style(regName, CYAN)}... `);
      try {
        const index = await refreshIndex(regName, entry, true);
        const pluginCount = index.plugins.length;
        print(`${style("done", GREEN)} (${pluginCount} plugin${pluginCount !== 1 ? "s" : ""})`);
      } catch (err) {
        print(`${style("failed", RED)} (${err instanceof Error ? err.message : String(err)})`);
      }
    }
  },
};
